client/blocks/login/two-factor-authentication/push-notification-approval-poller.jsx

```diff
@@ -1,42 +1,96 @@
-import PropTypes from 'prop-types';
-import { Component } from 'react';
-import { connect } from 'react-redux';
-import { startPollAppPushAuth, stopPollAppPushAuth } from 'calypso/state/login/actions';
-import { getTwoFactorPushPollSuccess } from 'calypso/state/login/selectors';
-
-class PushNotificationApprovalPoller extends Component {
-	static propTypes = {
-		onSuccess: PropTypes.func.isRequired,
-		pushSuccess: PropTypes.bool.isRequired,
-		startPollAppPushAuth: PropTypes.func.isRequired,
-		stopPollAppPushAuth: PropTypes.func.isRequired,
-	};
-
-	componentDidMount() {
-		this.props.startPollAppPushAuth();
-	}
+import config from '@automattic/calypso-config';
+import { useEffect, useRef } from 'react';
+import { useDispatch } from 'react-redux';
+import { localizeUrl } from 'calypso/lib/i18n-utils';
+import { updateNonce } from 'calypso/state/login/actions';
+import { remoteLoginUser } from 'calypso/state/login/actions/remote-login-user';
+import {
+	getTwoFactorAuthNonce,
+	getTwoFactorPushToken,
+	getTwoFactorUserId,
+} from 'calypso/state/login/selectors';
 
-	componentWillUnmount() {
-		this.props.stopPollAppPushAuth();
-	}
+const POLL_APP_PUSH_INTERVAL = 5 * 1000;
 
-	componentDidUpdate( prevProps ) {
-		if ( ! prevProps.pushSuccess && this.props.pushSuccess ) {
-			this.props.onSuccess();
-		}
-	}
+async function request( state ) {
+	const url = localizeUrl(
+		'https://wordpress.com/wp-login.php?action=two-step-authentication-endpoint'
+	);
+	const body = new URLSearchParams( {
+		user_id: getTwoFactorUserId( state ),
+		auth_type: 'push',
+		remember_me: true,
+		two_step_nonce: getTwoFactorAuthNonce( state, 'push' ),
+		two_step_push_token: getTwoFactorPushToken( state ),
+		client_id: config( 'wpcom_signup_id' ),
+		client_secret: config( 'wpcom_signup_key' ),
+	} );
 
-	render() {
-		return null;
-	}
+	const response = await fetch( url, {
+		method: 'POST',
+		credentials: 'include',
+		body,
+	} );
+
+	return await response.json();
 }
 
-export default connect(
-	( state ) => ( {
-		pushSuccess: getTwoFactorPushPollSuccess( state ),
-	} ),
-	{
-		startPollAppPushAuth,
-		stopPollAppPushAuth,
+const poll = ( signal ) => async ( dispatch, getState ) => {
+	let aborted = false;
+	signal.addEventListener( 'abort', () => {
+		aborted = true;
+	} );
+
+	while ( true ) {
+		try {
+			const response = await request( getState() );
+			if ( response.success ) {
+				const tokenLinks = response.data.token_links;
+				if ( Array.isArray( tokenLinks ) ) {
+					await remoteLoginUser( tokenLinks );
+				}
+				return true;
+			}
+
+			const twoStepNonce = response.data.two_step_nonce;
+			// if there is a `success: false` response without a nonce, that means
+			// we can't do the next iteration of the loop and we need to abort.
+			if ( ! twoStepNonce ) {
+				return false;
+			}
+
+			dispatch( updateNonce( 'push', twoStepNonce ) );
+		} catch {
+			// retry when the request fails for network-ish reasons
```

Comment:
Retrying is triggered from the poller, right? Should this be made clearer in the comment?

Reply:
No, retrying is triggered by the fact that we ignore the exception here and proceed to do another iteration of the `while ( true )` loop.

The `await request()` call will resolve to a `response` object even if the HTTP status is not 200. It can be (and is) 403 and there's still valid response JSON that we read.

But if the `fetch()` rejects, it's usually a network error. In that case we want to retry.

Should I explain that better in the comments?

Reply:
Yeah, I think that can be useful, this bit can take a while to understand if just reading at the code.

Reply:
Added more comments to the poll loop in 88a8e2269c.

Reply:
Thanks a lot!